contracts/IbcMiddleware.sol

```diff
@@ -0,0 +1,218 @@
+    ) external returns (AckPacket memory ackPacket);
+
+    function onRecvMWAck(
+        bytes32 channelId,
+        UniversalPacket calldata ucPacket,
+        // 0-based receiver middleware index in the MW stack.
+        //  0 for the first MW directly called by UniversalChannel MW.
+        // `mwIndex-1` is the last MW that delivers the packet to the non-MW dApp.
+        // Each mw in the stack must increment mwIndex by 1 before calling the next MW.
+        uint256 mwIndex,
+        address[] calldata mwAddrs,
+        AckPacket calldata ack
+    ) external;
+
+    function onRecvMWTimeout(
+        bytes32 channelId,
+        UniversalPacket calldata ucPacket,
+        uint256 mwIndex,
+        address[] calldata mwAddrs
+    ) external;
+}
+
+// dApps and IBC middleware contracts need to implement this interface to receive universal channel packets as packets' final destination.
+interface IbcUniversalPacketReceiver {
+    function onRecvUniversalPacket(bytes32 channelId, UniversalPacket calldata ucPacket)
+        external
+        returns (AckPacket memory ackPacket);
+
+    function onUniversalAcknowledgement(bytes32 channelId, UniversalPacket memory packet, AckPacket calldata ack)
+        external;
+
+    function onTimeoutUniversalPacket(bytes32 channelId, UniversalPacket calldata packet) external;
+}
+
+interface IbcMiddlwareProvider is IbcUniversalPacketSender, IbcMwPacketSender {
+    /**
+     * @dev MW_ID is the ID of MW contract on all supported virtual chains.
+     * MW_ID must:
+     * - be globally unique, ie. no two MWs should have the same MW_ID registered on Polymer chain.
+     * - be identical on all supported virtual chains.
+     * - be identical on one virtual chain across multiple deployed MW instances. Each MW instance belong exclusively to one MW stack.
+     * - be 1 << N, where N is a non-negative integer [0,255], and not in conflict with other MWs.
+     */
+    function MW_ID() external view returns (uint256);
+}
+
+/**
+ * @title IbcMiddleware
+ * @notice IBC middleware interface must be implemented by a IBC-middleware contract, similar to ICS-30
+ *  https://github.com/cosmos/ibc/tree/main/spec/app/ics-030-middleware.
+ * Current limitations:
+ *   - IbcMiddleware must sit on top of a UniversalChannel MW or another IbcMiddleware.
+ *   - IbcMiddleware cannnot own an IBC channel. Instead, UniversalChannel MW owns channels.
+ */
+interface IbcMiddleware is IbcMiddlwareProvider, IbcMwPacketReceiver, IbcUniversalPacketReceiver {}
+
+/**
+ * @title IbcUniversalChannelMW
+ * @notice This interface must be implemented by IBC universal channel middleware contracts.
+ * IbcUniversalChannelMW is a special type of IbcMiddleware that owns IBC channels, which are multiplexed for other IbcMiddleware.
+ * IbcUniversalChannelMW cannot sit on top of other MW, and must talk to IbcDispatcher directly.
+ */
+interface IbcUniversalChannelMW is IbcMiddlwareProvider, IbcPacketReceiver, IbcChannelReceiver {}
```

Comment:
Does this need to be a separate interface if only one contract impls it?

I guess for later versions?